api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java

```diff
@@ -111,53 +114,66 @@
-      traceFlags = TraceFlags.fromLowerBase16(traceparent, TRACE_OPTION_OFFSET);
+      contextFromParentHeader = extractContextFromTraceParent(traceparent);
     } catch (IllegalArgumentException e) {
-      throw new IllegalArgumentException("Invalid traceparent: " + traceparent, e);
+      logger.info("Unparseable traceparent header. Returning INVALID span context.");
+      return INVALID_SPAN_CONTEXT;
     }
 
-    String traceState = getter.get(carrier, TRACE_STATE);
+    String traceStateHeader = getter.get(carrier, TRACE_STATE);
     try {
-      if (traceState == null || traceState.isEmpty()) {
-        return SpanContext.createFromRemoteParent(traceId, spanId, traceFlags, TRACE_STATE_DEFAULT);
-      }
-      TraceState.Builder traceStateBuilder = TraceState.builder();
-      String[] listMembers = TRACESTATE_ENTRY_DELIMITER_SPLIT_PATTERN.split(traceState);
-      checkArgument(
-          listMembers.length <= TRACESTATE_MAX_MEMBERS, "TraceState has too many elements.");
-      // Iterate in reverse order because when call builder set the elements is added in the
-      // front of the list.
-      for (int i = listMembers.length - 1; i >= 0; i--) {
-        String listMember = listMembers[i];
-        int index = listMember.indexOf(TRACESTATE_KEY_VALUE_DELIMITER);
-        checkArgument(index != -1, "Invalid TraceState list-member format.");
-        traceStateBuilder.set(listMember.substring(0, index), listMember.substring(index + 1));
+      if (traceStateHeader == null || traceStateHeader.isEmpty()) {
+        return contextFromParentHeader;
       }
+      TraceState state = extractTraceState(traceStateHeader);
       return SpanContext.createFromRemoteParent(
-          traceId, spanId, traceFlags, traceStateBuilder.build());
+          contextFromParentHeader.getTraceId(),
+          contextFromParentHeader.getSpanId(),
+          contextFromParentHeader.getTraceFlags(),
+          state);
     } catch (IllegalArgumentException e) {
-      throw new IllegalArgumentException("Invalid tracestate: " + traceState, e);
+      logger.info("Unparseable tracestate header. Returning span context without state.");
+      return contextFromParentHeader;
+    }
+  }
+
+  private static SpanContext extractContextFromTraceParent(String traceparent) {
+    // TODO(bdrutu): Do we need to verify that version is hex and that
+    // for the version the length is the expected one?
+    checkArgument(
+        traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER
+            && (traceparent.length() == TRACEPARENT_HEADER_SIZE
+                || (traceparent.length() > TRACEPARENT_HEADER_SIZE
+                    && traceparent.charAt(TRACEPARENT_HEADER_SIZE) == TRACEPARENT_DELIMITER))
+            && traceparent.charAt(SPAN_ID_OFFSET - 1) == TRACEPARENT_DELIMITER
+            && traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER,
+        "Missing or malformed TRACEPARENT.");
```

Comment:
If we always return Invalid why do we create exceptions, maybe just simply return Invalid from here.

Reply:
It's a great question. And a great answer. I'll make it so.  :)

Reply:
cleaned this up a bit to not use exceptions where not currently necessary. We could get rid of some more exception handling by adding methods on the TraceId, SpanId to do format validation without exceptions.